netty-reactive-streams/src/main/java/com/typesafe/netty/HandlerPublisher.java

```diff
@@ -203,32 +203,40 @@ private void provideSubscriber(Subscriber<? super T> subscriber) {
 
     @Override
     public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
-        verifyRegisteredWithRightExecutor(ctx);
+        // If the channel is not yet registered, then it's not safe to invoke any methods on it, eg read() or close()
+        // So don't provide the context until it is registered.
+        if (ctx.channel().isRegistered()) {
+            provideChannelContext(ctx);
+        }
+    }
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+        provideChannelContext(ctx);
+        ctx.fireChannelRegistered();
+    }
 
+    private void provideChannelContext(ChannelHandlerContext ctx) {
         switch(state) {
             case NO_SUBSCRIBER_OR_CONTEXT:
+                verifyRegisteredWithRightExecutor(ctx);
                 this.ctx = ctx;
                 // It's set, we don't have a subscriber
                 state = NO_SUBSCRIBER;
                 break;
             case NO_CONTEXT:
+                verifyRegisteredWithRightExecutor(ctx);
                 this.ctx = ctx;
                 state = IDLE;
                 subscriber.onSubscribe(new ChannelSubscription());
                 break;
             default:
-                throw new IllegalStateException("This handler has already been placed in a handler pipeline");
+                // Ignore, this could be invoked twice by both handlerAdded and channelRegistered.
         }
     }
 
-    @Override
-    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
-        verifyRegisteredWithRightExecutor(ctx);
-        ctx.fireChannelRegistered();
-    }
-
     private void verifyRegisteredWithRightExecutor(ChannelHandlerContext ctx) {
```

Comment:
Can you just inline this method and put it at the top of `provideChannelContext(ctx)`? Having the code in one place would make it slightly easier to follow. (Just a minor thing.)

The method's code might be called more than once, but that should be OK, because it looks pretty cheap to execute.